### ISSUE TITLE
fp16 matmul collapses 23x from M=1 to M=2 on sm_75 without tensor cores (GTX 16-series); T4 and fp32 unaffected

### ISSUE THREAD

**Issue by LUOaini1213** [pytorch/pytorch#195716]
### 🐛 Describe the bug

On Turing parts **without tensor cores** (the GTX 16-series — `sm_75`, but with the
tensor cores removed), fp16 `matmul` loses ~23x of its effective bandwidth the moment M
goes from 1 to 2, and never recovers for larger M. fp32 at the same shapes shows no such
cliff, and a **tensor-core-equipped `sm_75` part (T4) shows no cliff either**.

M=1 (GEMV) runs near peak memory bandwidth; every M>=2 case (GEMM) runs at ~4% of it.
The weight-matrix traffic is identical in both cases, so this looks like kernel selection
for small-M fp16 GEMM on SIMT-only Turing rather than a fundamental limit.

### Controlled comparison

Both GPUs are `sm_75`, same architecture, same compute capability. The only difference
that matters here is that the T4 retains tensor cores and the GTX 1650 does not.

**fp16, `[M,896] @ [896,4864]`** (a real layer shape: Qwen2.5-0.5B MLP gate/up)

| M | GTX 1650 (no TC) | rel. M=1 | Tesla T4 (TC) | rel. M=1 |
|---|---|---|---|---|
| 1 | 99.5 GiB/s | 1.00x | 106.7 GiB/s | 1.00x |
| **2** | **4.3 GiB/s** | **0.04x** | **88.1 GiB/s** | **0.83x** |
| 4 | 4.5 GiB/s | 0.05x | 90.1 GiB/s | 0.84x |
| 8 | 4.5 GiB/s | 0.05x | 88.5 GiB/s | 0.83x |
| 16 | 4.4 GiB/s | 0.04x | 97.6 GiB/s | 0.91x |
| 64 | 4.5 GiB/s | 0.05x | 79.3 GiB/s | 0.74x |

Bandwidth is computed over the weight matrix (`K*N*itemsize`), which dominates traffic at
small M and keeps the two dtypes comparable.

Two further shapes from the same model reproduce the identical pattern on the GTX 1650
(M=1 -> M=2): `[M,4864]@[4864,896]` 74.9 -> 5.0 GiB/s (0.07x), `[M,896]@[896,1152]`
62.1 -> 4.6 GiB/s (0.07x). On the T4 the same shapes give 0.85x and 0.86x — no cliff.

**fp32 on the GTX 1650 has no cliff**, which is what makes this surprising: fp16 at M=2
takes 1.87 ms while fp32 at M=2 takes 0.138 ms for the identical shape — **13.5x slower in
wall time despite moving half the bytes**.

| M | fp32 GiB/s (GTX 1650) | rel. M=1 |
|---|---|---|
| 1 | 115.0 | 1.00x |
| 2 | 117.6 | 1.02x |
| 8 | 75.2 | 0.65x |
| 16 | 42.5 | 0.37x |

Interestingly the relationship inverts on the T4: there, fp32 degrades *faster* than fp16
at large M (M=64: 0.39x vs 0.74x), which is the expected behaviour when tensor cores are
available.

### Downstream impact

Serving Qwen2.5-0.5B with vLLM 0.28 on the GTX 1650, N requests issued simultaneously,
64 output tokens each, no further arrivals:

| concurrent seqs | fp16 tok/s | fp32 tok/s | fp32 advantage |
|---|---|---|---|
| 1 | 109.5 | 82.5 | 0.75x |
| 2 | 13.3 | 85.2 | **6.4x** |
| 4 | 22.2 | 203.1 | **9.1x** |
| 8 | 41.6 | 274.5 | **6.6x** |
| 16 | 80.5 | 295.4 | **3.7x** |

Under fp16, wall-clock for the whole burst was pinned at ~7.8 s for every batch size from
2 to 16 — a fixed per-step cost independent of batch. I ruled out CUDA graphs,
`max-num-seqs` padding, KV-cache sizing and chunked-prefill interleaving before isolating
it to `matmul`.

### Repro

Runs in a few seconds, no model download:

```python
import torch, statistics as st

SHAPES = [("MLP gate/up  [M,896]@[896,4864]", 896, 4864),
          ("MLP down     [M,4864]@[4864,896]", 4864, 896),
          ("QKV proj     [M,896]@[896,1152]", 896, 1152)]

def bench(M, K, N, dtype, iters=200):
    a = torch.randn(M, K, device="cuda", dtype=dtype)
    b = torch.randn(K, N, device="cuda", dtype=dtype)
    for _ in range(20): a @ b
    torch.cuda.synchronize()
    ts = []
    for _ in range(iters):
        s, e = torch.cuda.Event(True), torch.cuda.Event(True)
        s.record(); a @ b; e.record(); torch.cuda.synchronize()
        ts.append(s.elapsed_time(e))
    t = st.median(ts)
    return t, K * N * a.element_size() / t * 1e3 / 2**30

p = torch.cuda.get_device_properties(0)
print(f"GPU: {p.name} | sm_{p.major}{p.minor} | torch {torch.__version__}")
for dtype, tag in [(torch.float16, "fp16"), (torch.float32, "fp32")]:
    print(f"\n########## {tag} ##########")
    for name, K, N in SHAPES:
        print(f"\n{name}\n  {'M':>4}{'ms':>11}{'GiB/s':>10}{'rel M=1':>10}")
        base = None
        for M in [1, 2, 4, 8, 16, 32, 64]:
            t, gbs = bench(M, K, N, dtype)
            base = base or gbs
            print(f"  {M:>4}{t:>11.4f}{gbs:>10.1f}{gbs/base:>9.2f}x")
```

Full scripts and raw logs for both GPUs:
https://github.com/LUOaini1213/vllm-sm75-throughput

### Question

Is small-M fp16 GEMM on tensor-core-less `sm_75` expected to land on a path this much
slower than fp32, or is there a kernel-selection improvement available here? If it is
expected, it may be worth documenting — the natural assumption is that fp16 is never
dramatically slower than fp32 on the same hardware, and the failure mode is quiet
(no warning, just 20x less throughput).

### Versions

```
GTX 1650 (no tensor cores): torch 2.13.0+cu130, CUDA 13.0, driver 592.82,
    WSL2 Ubuntu 24.04.4, kernel 6.18.33.2-microsoft-standard-WSL2
Tesla T4 (tensor cores):    torch 2.11.0+cu128, CUDA 12.8, driver 580.82.07,
    Google Colab
```

cc @csarofeen @ptrblck @eqy @nWEIdia
